### PR TITLE
CAP - BLS12-381 host functions

### PR DESCRIPTION
**Pull request by jayz22**
Related discussions:
- https://github.com/stellar/rs-soroban-env/issues/779
- https://github.com/stellar/stellar-protocol/discussions/1500

### CODE REVIEW

File: core/cap-0059.md [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+
+## Simple Summary
+BLS12-381 is a pairing-friendly elliptic curve which enables a new suite of applications. This CAP proposes a set of new host functions providing access to BLS12-381 curve operations.
+
+## Working Group
+
+As described in the preamble section.
+
+## Motivation
+Pairing friendly elliptic curve operations are the backbone of many privacy-preserving applications, such as BLS signature, zkSNARKs, and anonymous login. BLS12-381 is one of the most well-known and widely adopted pairing friendly curves, due to its efficiency and 128-bit security. However, the curve operations are inherently computationally intensive, making them prohibitive to be implemented in a smart contract. Providing an efficient implementation for a well-chosen set of curve and field operations natively inside the Soroban host is crucial for unlocking privacy-preserving applications in smart contracts.
```
Comment (tomerweller) [inline core/cap-0059.md:24]:
```suggestion
Pairing friendly elliptic curve operations are the backbone of many advanced Zero Knowledge (ZK) constructions such as as zkSNARKs, zkSTARKs, PLONK, BLS Signatures and more. These, in turn, enable a wide array of applications ranging from scaling, to identity management and privacy. BLS12-381 is one of the most well-known and widely adopted pairing friendly curves, due to its efficiency and 128-bit security. However, the curve operations are inherently computationally intensive, making them prohibitive to be implemented in a smart contract. Providing an efficient implementation for a well-chosen set of curve and field operations natively inside the Soroban host is crucial for unlocking privacy-preserving applications in smart contracts.
```

**Review by tomerweller — CHANGES_REQUESTED** [review]
one nit

File: core/cap-0059.md [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+This CAP is aligned with the following Stellar Network Goals:
+
+  - The Stellar Network should make it easy for developers of Stellar projects
+    to create highly usable products
+
+## Abstract
+11 new host functions are proposed for performing field and curve operations on BLS12-381. Definitions and the semantics of each new host function, as well as the associated new metering parameters, will be introduced and explained. 
+
+## Specification
+
```
Comment (anupsdf) [inline core/cap-0059.md:36]: How about specifying the length and endianness of the BytesObject?

Comment (anupsdf) [inline core/cap-0059.md:36]: Should we specify whether points on the curves used in the host functions are in uncompressed(x and y coordinates) or compressed (x coordinate) form? 
From XDR changes it looks like we are using uncompressed form.

File: core/cap-0059.md [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+### New host functions
```
Comment (anupsdf) [inline core/cap-0059.md:37]: How did we come up with the list of host functions to add? Is there a mapping of use-case to host function somewhere? That way we know which use cases will work and if any host function will never be used.

File: core/cap-0059.md (hunk 2 above)
Comment (jayz22) [reply core/cap-0059.md:36]: done

Comment (jayz22) [reply core/cap-0059.md:36]: done

File: core/cap-0059.md (hunk 3 above)
Comment (jayz22) [reply core/cap-0059.md:37]: It was mostly discussed in https://github.com/stellar/rs-soroban-env/issues/779 
and I've also mentioned in the design rationale.

**Review by tomerweller — DISMISSED** [review]

**Review by tomerweller — APPROVED** [review]

**Pull request merged**